# MATH 210 Introduction to Mathematical Computing

**February 2, 2026**

* Roots and IVT
* Bisection Method
* Examples

## Roots and IVT

A **root** of a function $g(x)$ is a value $r$ such that $g(r) = 0$. We also call a root a **zero** of the function.

What's the connection between fixed points and roots? If $r$ is a root of $g(x)$ then $r$ is a fixed point of $f(x) = x - g(x)$.

Recall the Intermediate Value Theorem:

Let $g(x)$ be continuous on $[a,b]$. If $g(a)g(b) < 0$ then there is some $c \in (a,b)$ such that $g(c) = 0$.

## Bisection Method

The bisection method is a root finding method. It applies IVT over and over again taking smaller and smaller intervals.

* Start with $[a_0,b_0]$ such that $f(a_0) f(b_0) <0$.
* Let $m_0 = (a_0 + b_0)/2$ be the midpoint.
* Then either: (i) $f(a_0)f(m_0) < 0$ or (ii) $f(m_0)f(b_0) < 0$ (or $f(m_0) = 0$).
* Define: (i) $a_1 = a_0$, $b_1 = m_0$ or (ii) $a_1 = m_0$, $b_1 = b_0$ so that $r \in (a_1,b_1)$.
* Repeat!

Write a function called `bisection` which takes function `f`, numbers `a` and `b` and integer `N`, and return the midpoint $m_N$ after $N$ iterations of the bisection mtehod applied to $f(x)$.

In [26]:
def bisection(f,a,b,N):
    if f(a)*f(b) >= 0:
        print("Initial interval does not satisfy f(a)f(b) < 0.")
        return None
    a_n = a
    b_n = b
    for n in range(1,N+1):
        m_n = (a_n + b_n)/2
        f_m_n = f(m_n)
        if f(a_n)*f_m_n < 0:
            b_n = m_n
        elif f(b_n)*f_m_n < 0:
            a_n = m_n
        elif f_m_n == 0:
            print("Found exact solution.")
            return m_n
        else:
            print("Bisection method fails.")
            return None
    return (a_n + b_n)/2

In [27]:
f = lambda x: x**3 + x - 1
a = 0
b = 1
N = 10
r = bisection(f,a,b,N)
print(r)

0.68212890625


## Example

Find fixed points of $f(x) = \sqrt{1 + x} + \sqrt{x}$ using:

* Bisection method
* Fixed point iteration

First, how many fixed points are there? There is 1. Prove it!

Now, translate fixed point of $f(x)$ to root of $g(x) = f(x) - x$.

In [28]:
g = lambda x: (1 + x)**0.5 + x**0.5 - x

In [29]:
print(g(4),g(5))

0.2360679774997898 -0.3144422797170323


In [31]:
a = 4
b = 5
N = 10
r = bisection(g,a,b,N)
print(r)

4.43896484375


In [32]:
g(r)

7.909838433128868e-05

Apply fixed point iteration to find the fixed point.

In [33]:
f = lambda x: (1 + x)**0.5 + x**0.5

In [34]:
f(r)

4.439043942134331

In [35]:
x = 4.5
N = 50
for n in range(1,N+1):
    x = f(x)
print(x)

4.439109106851355


It looks like fixed point iteration is converging to $L \approx 4.439$. Prove $|f'(x)| < 1$ for $x$ near $L$.

$$
f'(x) = \frac{1}{2 \sqrt{1 + x}} + \frac{1}{2\sqrt{x}}
$$

Since each denominator is increasing as $x$ is increasing, the value of each term is decreasing as $x$ is increasing.

Look at $f'(1) = \frac{1}{2 \sqrt{2}} + \frac{1}{2} < 1$ and note $f'(x)$ is always positive and decreasing for $x$ increasing from 1. Therefore $|f'(x)| < 1$ for all $x \geq 1$ and so fixed point iteration converges for any $x \geq 1$.

In [36]:
x = 1000
N = 50
for n in range(1,N+1):
    x = f(x)
print(x)

4.439109106851355
